From: Christian Marangi <ansuelsmth@gmail.com>
Date: Mon, 14 Apr 2025 18:04:25 +0200
Subject: [PATCH 1/2] mm: permit to declare custom execmem alloc/free function

Permit to declare custom execmem alloc/free function that bypass the
execmem API. This works by making the alloc/free function weak
permitting an arch to declare a replacement for them.

Signed-off-by: Christian Marangi <ansuelsmth@gmail.com>
Co-authored-by: Shiji Yang <yangshiji66@outlook.com>
---
 include/linux/moduleloader.h |  4 ++++
 mm/execmem.c                 | 14 ++++++++++++--
 2 files changed, 16 insertions(+), 2 deletions(-)

--- a/include/linux/moduleloader.h
+++ b/include/linux/moduleloader.h
@@ -122,4 +122,8 @@
 /* Any cleanup before freeing mod->module_init */
 void module_arch_freeing_init(struct module *mod);
 
+enum execmem_type;
+void *arch_execmem_alloc(enum execmem_type type, size_t size);
+void arch_execmem_free(void *ptr);
+
 #endif
--- a/mm/execmem.c
+++ b/mm/execmem.c
@@ -461,7 +461,7 @@
 }
 #endif /* CONFIG_ARCH_HAS_EXECMEM_ROX */
 
-void *execmem_alloc(enum execmem_type type, size_t size)
+void *__weak arch_execmem_alloc(enum execmem_type type, size_t size)
 {
 	struct execmem_range *range = &execmem_info->ranges[type];
 	bool use_cache = range->flags & EXECMEM_ROX_CACHE;
@@ -494,7 +494,12 @@
 	return no_free_ptr(p);
 }
 
-void execmem_free(void *ptr)
+void *execmem_alloc(enum execmem_type type, size_t size)
+{
+	return arch_execmem_alloc(type, size);
+}
+
+void __weak arch_execmem_free(void *ptr)
 {
 	/*
 	 * This memory may be RO, and freeing RO memory in an interrupt is not
@@ -511,6 +516,11 @@
 	return !!(execmem_info->ranges[type].flags & EXECMEM_ROX_CACHE);
 }
 
+void execmem_free(void *ptr)
+{
+	arch_execmem_free(ptr);
+}
+
 static bool execmem_validate(struct execmem_info *info)
 {
 	struct execmem_range *r = &info->ranges[EXECMEM_DEFAULT];
